nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp

```diff
@@ -112,7 +112,18 @@ class BtActionNode : public BT::CoroActionNode
     on_tick();
 
 new_goal_received:
-    auto future_goal_handle = action_client_->async_send_goal(goal_);
+    goal_result_available_ = false;
+    auto send_goal_options = typename rclcpp_action::Client<ActionT>::SendGoalOptions();
+    send_goal_options.result_callback =
+      [this](const typename rclcpp_action::ClientGoalHandle<ActionT>::WrappedResult & result) {
+        if (result.code != rclcpp_action::ResultCode::ABORTED) {
+          goal_result_available_ = true;
+          result_ = result;
+        }
```

Comment:
is this callback ensured to be run atomically with respect to that polling loop down below because of that

`      rclcpp::spin_some(node_);
` at line 151?

Reply:
This callback doesn't trigger anything in particular, it just sets some variables for the next iteration. I don't think that creates any async issues as the cases of results are:
- Abort -- which are ignored. This is where async issues could occur but for that same reason we dont use them
- Success / fail / any other status -- are set for the next loop cycle

Reply:
Sorry - reread the question, I think what you mean is the spin going to be run such that the variables need to be locked with a mutex or something -- no I dont think so. Since spin will execute the callbacks and block until complete and then allow for it to continue. The callbacks won't run at the same time as the BT execution as long as no one does a blocking spin thread on that same node (which no one does, because if they did, then this would have worked without it, which it didnt).

Reply:
excellent, this is what I meant